py-tensorflow: add versions 2.5.0 and 2.6.0

glennpj opened the pull request:
This PR closes #22481, from which some things were pulled for this PR.
- add version 2.5.0
- add version 2.6.0
- add patches for newer protobuf
- set constraints
- fixes #28612

glennpj committed:
py-tensorflow: add versions 2.5.0 and 2.6.0

- add version 2.5.0
- add version 2.6.0
- add patches for newer protobuf
- set constraints

glennpj committed:
Remove import os. left over from testing

glennpj committed:
Remove unused patch file

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -111,7 +113,8 @@ class PyTensorflow(Package, CudaPackage):
     # See _TF_MIN_BAZEL_VERSION and _TF_MAX_BAZEL_VERSION in configure.py
-    depends_on('bazel@3.1.0:3',  type='build', when='@2.3:')
+    depends_on('bazel@3.7.2:3.99.0',  type='build', when='@2.5:')
+    depends_on('bazel@3.1.0:3.99.0',  type='build', when='@2.3:2.4.999')
```
```suggestion
    depends_on('bazel@3.1.0:3.99.0',  type='build', when='@2.3:2.4')
```

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -148,16 +151,19 @@ class PyTensorflow(Package, CudaPackage):
     depends_on('py-backports-weakref@1.0rc1', type=('build', 'run'), when='@1.2.0:1.2.1')
```
You're missing a dep on `py-clang`, which you'll need to create a package for.

glennpj replied:
I just added this. It is somewhat surprising that this dependency is for clang-5.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -148,16 +151,19 @@ class PyTensorflow(Package, CudaPackage):
     # propagate the mpi variant setting for h5py/hdf5 to avoid unexpected crashes
-    depends_on('py-h5py@2.10.0:2.10+mpi', type=('build', 'run'), when='@2.2:+mpi')
-    depends_on('py-h5py@2.10.0:2.10~mpi', type=('build', 'run'), when='@2.2:~mpi')
+    depends_on('py-h5py@3.1.0:3.1.999', type=('build', 'run'), when='@2.5:+mpi')
+    depends_on('py-h5py~mpi@3.1.0:3.1.999', type=('build', 'run'), when='@2.5:~mpi')
+    depends_on('py-h5py@2.10.0:2.10.999', type=('build', 'run'), when='@2.2:2.4.999+mpi')
+    depends_on('py-h5py~mpi@2.10.0:2.10.999', type=('build', 'run'), when='@2.2:2.4.999~mpi')
```
```suggestion
    depends_on('py-h5py@3.1.0:3.1', type=('build', 'run'), when='@2.5:+mpi')
    depends_on('py-h5py~mpi@3.1.0:3.1', type=('build', 'run'), when='@2.5:~mpi')
    depends_on('py-h5py@2.10.0:2.10', type=('build', 'run'), when='@2.2:2.4.999+mpi')
    depends_on('py-h5py~mpi@2.10.0:2.10', type=('build', 'run'), when='@2.2:2.4.999~mpi')
```

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -148,16 +151,19 @@ class PyTensorflow(Package, CudaPackage):
     depends_on('py-enum34@1.1.6:', type=('build', 'run'), when='@1.4.0:1.4.1')
-    depends_on('py-gast@0.3.3', type=('build', 'run'), when='@2.2:')
+    depends_on('py-gast@0.4.0', type=('build', 'run'), when='@2.5:')
+    depends_on('py-gast@0.3.3', type=('build', 'run'), when='@2.2:2.4.999')
```
```suggestion
    depends_on('py-gast@0.3.3', type=('build', 'run'), when='@2.2:2.4')
```

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -216,15 +222,17 @@ class PyTensorflow(Package, CudaPackage):
     # depends_on('trisycl',    when='+opencl~computepp')
     depends_on('cuda@:10.2', when='+cuda @:2.3')
-    depends_on('cuda@:11.1', when='+cuda @2.4.0:')
+    depends_on('cuda@:11.1', when='+cuda @2.4.0:2.4.999')
```
```suggestion
    depends_on('cuda@:11.1', when='+cuda @2.4.0:2.4')
```

glennpj replied:
Hmm, can we just state, `@2.4` there?

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -216,15 +222,17 @@ class PyTensorflow(Package, CudaPackage):
         # Only builds correctly on little-endian machines
         depends_on('py-grpcio@1.8.6:', type=('build', 'run'), when='@1.8:2.3')
-        depends_on('py-grpcio@1.32.0:1.32', type=('build', 'run'), when='@2.4:')
+        depends_on('py-grpcio@1.32.0:1.32.999', type=('build', 'run'), when='@2.4:2.4.999')
+        depends_on('py-grpcio@1.34.0:1.34.999', type=('build', 'run'), when='@2.5:')
```
```suggestion
        depends_on('py-grpcio@1.32.0:1.32', type=('build', 'run'), when='@2.4.0:2.4')
        depends_on('py-grpcio@1.34.0:1.34', type=('build', 'run'), when='@2.5.0:2.5')
        depends_on('py-grpcio@1.37.0:1', type=('build', 'run'), when='@2.6.0:')
```

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -301,6 +309,20 @@ class PyTensorflow(Package, CudaPackage):
+    # needed for protobuf-3.16 and greater
+    patch('example_parsing.patch', when='^protobuf@3.16:')
+
+    # allow linker to befound in PATH
```
```suggestion
    # allow linker to be found in PATH
```

aweits commented:
You'll have to look into the keras dependency as well...

glennpj commented:
While I am looking into additional dependencies, what should be done about `py-tensorboard` and `py-tensorflow-estimator`? Is there a reason those have been left out of previous recipes?

glennpj committed:
Update var/spack/repos/builtin/packages/py-tensorflow/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

glennpj committed:
Update var/spack/repos/builtin/packages/py-tensorflow/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

glennpj committed:
Update var/spack/repos/builtin/packages/py-tensorflow/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

glennpj committed:
Update var/spack/repos/builtin/packages/py-tensorflow/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

glennpj committed:
Update var/spack/repos/builtin/packages/py-tensorflow/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

glennpj committed:
Update var/spack/repos/builtin/packages/py-tensorflow/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

adamjstewart commented:
I believe those were left out to prevent circular dependencies. We might need to actually comment them out in a patch.

aweits commented:
> While I am looking into additional dependencies, what should be done about `py-tensorboard` and `py-tensorflow-estimator`? Is there a reason those have been left out of previous recipes?

There are (or were) circular dependency issues that needed to be avoided. Pretty sure they're still issues.

glennpj committed:
Add py-clang dependency

glennpj commented:
> You'll have to look into the keras dependency as well...

This is complicated. The only 2.5 version available on PyPI is 2.5.0rc0 but only as a wheel. For 2.6.0 there is only a wheel available on PyPI but the source can be gotten from github, [Release Keras Release 2.6.0 · keras-team/keras](https://github.com/keras-team/keras/releases/tag/v2.6.0). I do not see a source release for 2.5.

glennpj committed:
Adjust py-clang constraint

glennpj commented:
This is from the Keras README:
```
Installation
Keras comes packaged with TensorFlow 2 as tensorflow.keras. To start using Keras, simply install TensorFlow 2.
```
Do we actually need a py-keras dependency?

adamjstewart commented on var/spack/repos/builtin/packages/py-clang/package.py:
```diff
@@ -0,0 +1,26 @@
+    """Python bindings for clang from clang release branches"""
+
+    homepage = "https://clang.llvm.org/"
+    pypi     = "clang/clang-5-5.0.tar.gz"
```
```suggestion
    pypi     = "clang/clang-5.0.tar.gz"
```

glennpj replied:
That was a carryover from https://pypi.org/project/clang-5/5.0/. That was the dependency in the spec but then I switched to the main clang project and pulled version 5.

adamjstewart commented on var/spack/repos/builtin/packages/py-clang/package.py:
```diff
@@ -0,0 +1,26 @@
+    version('6.0.0',   sha256='0ca2b8222156bb82b3bba060709f2b0159da99bef2bce346253719c740df20fa')
+    version('5.0',     sha256='ceccae97eda0225a5b44d42ffd61102e248325c2865ca53e4407746464a5333a')
+
+    depends_on('python@2.7:2.8,3.4:', type=('build', 'run'))
```
Where do you see that Python 2 is supported?

glennpj replied:
Version 5 can use python-2. I was focused on what was needed for tensorflow so this is not complete yet. I need to make another pass and work out the constraints as the versions change. Sorry, I should have specified that.

glennpj replied:
There are some inconsistencies in the dependency spec across versions but I adjusted these to what should be reasonable.

glennpj committed:
Update var/spack/repos/builtin/packages/py-clang/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

glennpj commented:
It looks to me like `py-tensorflow-estimator-2.6` needs `py-keras-2.6`, which now needs bazel.

glennpj commented:
I marked this PR as "draft" because this has snowballed into much more work with `keras`, `tensorboard` and `estimator`.

glennpj committed:
Build tensorflow with tensorboard

- tensorflow
    - added 2.6.1 and 2.6.2 versions

- tensorboard
    - have bazel use number of jobs set by spack
    - add versions and constraints

- new package: py-tensorboard-data-server
- use wheel for py-tensorboard-plugin-wit
  This package can not build with newer versions of bazel that are
  needed for newer versions of py-tensorboard.

glennpj committed:
Remove empty line at end of file

glennpj committed:
Fix import sorting

glennpj commented:
Finally getting back to this. Keras built in to tensorflow-2 is recommended for use as opposed to the standalone keras going forward. The same is true for tensorflow-estimator, which is built in to tensorflow-2. It does look like a new package is needed to get estimator working though, tensorflow_datasets.

aweits commented:
> Finally getting back to this. Keras built in to tensorflow-2 is recommended for use as opposed to the standalone keras going forward. The same is true for tensorflow-estimator, which is built in to tensorflow-2. It does look like a new package is needed to get estimator working though, tensorflow_datasets.

I'm a bit confused? - from RELEASE.md:
```
*   `tf.keras`:

    *   Keras been split into a separate PIP package (`keras`), and its code has
        been moved to the GitHub
        repository[keras-team/keras](http://github.com/keras-team/keras). The
        API endpoints for `tf.keras` stay unchanged, but are now backed by the
        `keras` PIP package. The existing code in tensorflow/python/keras is a
        staled copy and will be removed in future release (2.7). Please remove
        any imports to `tensorflow.python.keras` and replace them with public
        tf.keras API instead.
```

glennpj commented:
@aweits I am also confused by the way tensorflow is structured, but here is my understanding. There are effectively two keras APIs, `keras` and `tf.keras`. My understanding from the documentation and tutorials that I have been hitting is that `tf.keras` is preferred, and is current in tensorflow. The `tensorflow.python.keras` referred to is the other API and is the one that is being removed. I believe the split out of the `tf.keras` API is to allow for providing separate pip packages on PyPI, and probably conda. Now, my confidence level in my understanding of that is not as high as I would like but that is what I have determined to this point.

glennpj commented:
While tensorflow-2.6.2 is listed as the latest release, there is a 2.7.0 release on github. I am going to pull that one and try to get a little more clarity regarding keras.

aweits commented:
> While tensorflow-2.6.2 is listed as the latest release, there is a 2.7.0 release on github. I am going to pull that one and try to get a little more clarity regarding keras.

I think I may have started updating the keras package a while ago - I'll look to see if I can find that branch. I can exert some more effort on this shortly as I will have a need for this when I post the alphafold stuff I've been working on...

glennpj committed:
Adjust python dependencies on py-clang

glennpj commented:
I am not sure how much version 2.7.0 clarified anything. Looking at the release notes, there is definitely activity with regard to tf.keras.

[Release TensorFlow 2.7.0 · tensorflow/tensorflow](https://github.com/tensorflow/tensorflow/releases/tag/v2.7.0)

However, the code in `tensorflow/keras/python` is still present. I ran through some keras tutorials and verified that it does work. I will go ahead and add py-tensorboard-2.7.0 before I forget.

glennpj committed:
Add version 2.7.0 of pt-tensorflow and py-tensorboard

glennpj committed:
Adjust bazel constraints

glennpj commented:
I added tensorflow/tensorboard 2.7.0. Beginning with 2.7.0, bazel-4 can be used so I adjusted those constraints. Should I take this out of "draft" status?

aweits commented:
> I added tensorflow/tensorboard 2.7.0. Beginning with 2.6.0, bazel-4 can be used so I adjusted those constraints. Should I take this out of "draft" status?

Just tried to build 2.6.2 and it's insisting on bazel < 3.99.0. (The git history on this is confusing as they apparently changed things there and reverted at least once...)

glennpj commented:
> Just tried to build 2.6.2 and it's insisting on bazel < 3.99.0.

Argh! I tried 2.6.0 and 2.7.0, and just assumed that 2.6.1 and 2.6.2 would be the same.

glennpj commented:
I may have conflated that with tensorboard, which can use bazel-4. I am checking again.

glennpj committed:
bazel-4 support begins with py-tensorflow-2.7.0

glennpj commented:
Indeed, bazel-4 support begins with 2.7.0, not 2.6.0.

aweits commented:
FYI, currently doing some test builds with horovod against this...

aweits commented:
> FYI, currently doing some test builds with horovod against this...

looks like we do need the external keras. (I usually test with the horovod tensorflow2 synthetic benchmarks BTW)

```
[aweits@skl-a-00 tensorflow2]$ python ./tensorflow2_synthetic_benchmark.py
Traceback (most recent call last):
  File "./tensorflow2_synthetic_benchmark.py", line 21, in <module>
    import horovod.tensorflow as hvd
  File "/.autofs/scratch/aweits/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/py-horovod-master-l4kotmg2rh2emuoxdxnykjyi6e5wvjy5/lib/python3.8/site-packages/horovod/tensorflow/__init__.py", line 42, in <module>
    from horovod.tensorflow.sync_batch_norm import SyncBatchNormalization
  File "/.autofs/scratch/aweits/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/py-horovod-master-l4kotmg2rh2emuoxdxnykjyi6e5wvjy5/lib/python3.8/site-packages/horovod/tensorflow/sync_batch_norm.py", line 22, in <module>
    class SyncBatchNormalization(tf.keras.layers.BatchNormalization):
  File "/.autofs/scratch/aweits/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/py-tensorflow-2.6.2-7h65ftyq5frxkc6dow7h5t2hczzsxxyt/lib/python3.8/site-packages/tensorflow/python/util/lazy_loader.py", line 62, in __getattr__
    module = self._load()
  File "/.autofs/scratch/aweits/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/py-tensorflow-2.6.2-7h65ftyq5frxkc6dow7h5t2hczzsxxyt/lib/python3.8/site-packages/tensorflow/python/util/lazy_loader.py", line 45, in _load
    module = importlib.import_module(self.__name__)
  File "/.autofs/scratch/aweits/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/python-3.8.7-5ldw5aygmtku3dnhmulmds73dpglpe3c/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
ModuleNotFoundError: No module named 'keras'


```

glennpj commented:
That might be fixable by replacing `import keras` with `import tensorflow.keras as keras`. In any event, I am working on a stand-alone `py-keras` but have not been able to get it to build yet. Also, just a note that there are further dependency constraint tweaks needed for `py-tensorflow@2.7.0`. It seems dependency specs are up and down as versions change.

glennpj commented:
The problem that I am having with building py-keras is with protobuf. @aweits solved that in #27991. 

Even though the tensorflow setup.py states a dependency on keras, and keras has a dependency on tensorflow, I think the dependency should only be specified in the py-keras package, to avoid having a circular dependency.

glennpj committed:
Adjust dependencies

glennpj committed:
Loosen cuda constraint on versions > 2.5

Tensorflow-2.5 and above can use cuda up to version 11.4.

glennpj committed:
Add constraints to patch

The 0008-Fix-protobuf-errors-when-using-system-protobuf.patch patch
should only apply to versions 2.5 and above.

glennpj commented:
It looks like additional version constraints for protobuf are needed for older versions. Looking into that now.

glennpj committed:
Adjust constraints

- versions 2.4 and below need protobuf-3.12 and below
- versions 2.4 and above can use up to cuda-11.4
- versions 2.2 and below can not use cudnn-8
- the null_linker_bin patch should only be applied to versions 2.5 and
  above.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard-data-server/package.py:
```diff
@@ -0,0 +1,49 @@
+        pip = which('pip')
+        wheel_files_pattern = '*.whl'
+        wheel_files = glob.glob(wheel_files_pattern)
+        pip('install', '--prefix={0}'.format(prefix), *wheel_files)
```
Remind me to update this in #27798 after this is merged

adamjstewart replied:
#27798 has been merged now, you might be able to use the default `install` method, or you might need to modify it a bit.

adamjstewart replied:
Reminder to remove this install method, it should work with the default install method now

glennpj replied:
Removing the install method does not properly install the package. This is what is installed if the install method is removed:
```console
$ ls -R opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n
opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n:
lib

opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n/lib:
python3.9

opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n/lib/python3.9:
site-packages

opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n/lib/python3.9/site-packages:
UNKNOWN-0.0.0.dist-info

opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n/lib/python3.9/site-packages/UNKNOWN-0.0.0.dist-info:
AUTHORS  direct_url.json  INSTALLER  LICENSE  METADATA  RECORD  REQUESTED  top_level.txt  WHEEL
```
Here is what should be installed:
```console
$ ls -R opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n
opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n:
lib

opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n/lib:
python3.9

opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n/lib/python3.9:
site-packages

opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n/lib/python3.9/site-packages:
tensorboard_data_server  tensorboard_data_server-0.6.1.dist-info

opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n/lib/python3.9/site-packages/tensorboard_data_server:
bin  __init__.py  __pycache__

opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n/lib/python3.9/site-packages/tensorboard_data_server/bin:
server

opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n/lib/python3.9/site-packages/tensorboard_data_server/__pycache__:
__init__.cpython-39.pyc

opt/spack/linux-centos7-cascadelake/gcc-10.3.0/py-tensorboard-data-server-0.6.1-3ykcchuc6y2irnoascobwduy34ohab4n/lib/python3.9/site-packages/tensorboard_data_server-0.6.1.dist-info:
direct_url.json  INSTALLER  METADATA  RECORD  REQUESTED  top_level.txt  WHEEL
```

adamjstewart commented on var/spack/repos/builtin/packages/py-clang/package.py:
```diff
@@ -0,0 +1,27 @@
+
+    depends_on('python@2.7:2.8,3:', type=('build', 'run'), when='@:7')
+    depends_on('python@3:', type=('build', 'run'), when='@8:')
+    depends_on('py-setuptools', type='build')
```
I think LLVM is needed as a run-time dep

glennpj replied:
This has been added.

glennpj replied:
This introduces a problem. The spec for tensorflow-2.6 is:
```
'clang ~= 5.0'
```
That is a really old version of llvm, which means this will prevent tensorflow-2.6 from using modern versions of GCC when spack concretizes the spec.

glennpj replied:
I think there are a couple of ways forward:

1. drop the run dependency on llvm in py-clang, but have it in py-libclang
In addition, drop all versions except 5 and add comments to explain the purpose of the package and that py-libclang should be preferred. Whoever needs the bindings would need to ensure a compatible LLVM installation.
1. drop clang support from tensorflow-2.6
Remove the py-clang package and mask the clang dependency in setup.py. Support for clang would begin with tensorflow-2.7.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard-plugin-wit/package.py:
```diff
@@ -22,15 +22,23 @@ class PyTensorboardPluginWit(Package):
 
     version('master', branch='master')
-    version('1.8.0', sha256='1e4de1bbf6ae61c4d27b114ec2e1093bc4765b8c2bbb2cc5d43e2075b08a5fea')
+    version('1.8.0-py3',
```
Since there's no Python 2 version, let's just call this 1.8.0.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard-plugin-wit/package.py:
```diff
@@ -47,6 +55,10 @@ def setup_build_environment(self, env):
         env.set('TEST_TMPDIR', self.tmp_path)
 
+    @when('@1.8.0-py3')
+    def setup_build_environment(self, env):
```
If you instead add `@when('@1.7.0')` to the function above, you won't need this function.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard-plugin-wit/package.py:
```diff
@@ -69,6 +81,10 @@ def setup(self, spec, prefix):
                     '{0} setup.py'.format(spec['python'].command.path),
                     'tensorboard_plugin_wit/pip_package/build_pip_package.sh')
 
+    @when('@1.8.0-py3')
```
Same here

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard/package.py:
```diff
@@ -16,13 +16,17 @@ class PyTensorboard(Package):
     version('2.3.0', sha256='947a58702c2841eb4559637dbf8639633f79de9a0f422be9737f3563a1725440')
     version('2.2.0', sha256='d0dfbf0e4b3b5ebbc3fafa6d281d4b9aa5478eac6bac3330652ab6674278ab77')
 
     depends_on('python@2.7:2.8,3.2:', type=('build', 'run'))
```
Latest version requires Python 3.6+

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard/package.py:
```diff
@@ -33,19 +37,33 @@ class PyTensorboard(Package):
     depends_on('py-google-auth@1.6.3:1', type=('build', 'run'))
```
In the latest version it is `@1.6.3:2`

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard/package.py:
```diff
@@ -33,19 +37,33 @@ class PyTensorboard(Package):
+    # what is needed for py-tensorboard-2.5 and greater. Work around that for
+    # now by using a wheel.
+    depends_on('py-tensorboard-plugin-wit@1.6.0:', type=('build', 'run'), when='@2.2.0:2.4')
+    depends_on('py-tensorboard-plugin-wit@1.8.0-py3', type=('build', 'run'), when='@2.5.0:')
```
Version 2.7.0 doesn't require 1.8+, if Spack has concretization issues with 1.6 it will choose 1.8 instead. I would remove this and let the concretizer handle it. We could also convert 1.6 to a wheel if we need to.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -111,7 +116,8 @@ class PyTensorflow(Package, CudaPackage):
 
     # See _TF_MIN_BAZEL_VERSION and _TF_MAX_BAZEL_VERSION in configure.py
-    depends_on('bazel@3.1.0:3',  type='build', when='@2.3:')
+    depends_on('bazel@3.1.0:',        type='build', when='@2.7:')
```
Isn't it `3.7.2:4.99.0`?

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -138,38 +144,51 @@ class PyTensorflow(Package, CudaPackage):
 
     # Listed under REQUIRED_PACKAGES in tensorflow/tools/pip_package/setup.py
-    depends_on('py-absl-py@0.10:0', type=('build', 'run'), when='@2.4.0:')
+    depends_on('py-absl-py@0.4:', type=('build', 'run'), when='@2.7:')
```
I'm kind of shocked that so many min dependency versions were decreased instead of increased

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -138,38 +144,51 @@ class PyTensorflow(Package, CudaPackage):
     depends_on('py-astor@0.6.0:', type=('build', 'run'), when='@1.6:2.1')
     depends_on('py-backports-weakref@1.0:', type=('build', 'run'), when='@1.3: ^python@:3.3')
     depends_on('py-backports-weakref@1.0rc1', type=('build', 'run'), when='@1.2.0:1.2.1')
+    depends_on('py-clang@9.0.1:', type=('build', 'run'), when='@2.7:')
```
This should be `py-libclang`

glennpj replied:
Well, this is interesting. For tensorflow-2.6:
```
'clang ~= 5.0'
```
For tensorflow-2.7:
```
'libclang >= 9.0.1'
```
They are two different projects on PyPI, although I think they are compatible for common versions. Since I started this with tensorflow-2.6, I used the clang project. It runs from llvm-5 to llvm-11. The libclang project runs from llvm-9 to llvm-12. 

I could create a new py-libclang package, but that means that py-clang would only exist as a dependency for py-tensorflow-2.6. I do not see another way as libclang would not meet the requirements for tensorflow-2.6. I am worried about confusion between py-clang and py-libclang though. 

Perhaps the best option is to limit the versions of py-clang from 5 to 8, and have py-libclang from 9 and up, eliminating the crossover between versions 9 to 11. Perhaps there could be a virtual py-clang package. 

----
https://pypi.org/project/clang/
https://pypi.org/project/libclang/

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -138,38 +144,51 @@ class PyTensorflow(Package, CudaPackage):
     # propagate the mpi variant setting for h5py/hdf5 to avoid unexpected crashes
-    depends_on('py-h5py@2.10.0:2.10+mpi', type=('build', 'run'), when='@2.2:+mpi')
-    depends_on('py-h5py@2.10.0:2.10~mpi', type=('build', 'run'), when='@2.2:~mpi')
+    depends_on('py-h5py@2.9.0:3.1', type=('build', 'run'), when='@2.7:')
```
Latest version should be `2.9.0:`

glennpj replied:
The comment in the `setup.py` file is:
```
'h5py >= 2.9.0', # capped since 3.3.0 lacks py3.6
```
I interpreted that to mean that there was a cap around version 3.3, even though it is not explicit as it is for other capped specs. I can not recall now why I picked the cap to be 3.1 instead of 3.2 but I think the intent is for it to be capped. I certainly could be misinterpreting that though.

adamjstewart replied:
I'm wondering if the cap used to be there, but was removed for some reason and the comment forgotten. Would have to dig through git blame.

adamjstewart replied:
In general I try to keep the Spack recipe as close to possible as the `setup.py` recipe and only change things when I find that they don't build properly.

glennpj replied:
I do as well. I know why I selected 3.1 as the cap. It was because that is what it was for tensorflow-2.6. That was the most recent value that I could find so I kept it. 

For tensorflow-2.6:
```
'h5py ~= 3.1.0'
```

Perhaps the comment is referring to why it was capped for earlier versions, although that would be odd from a context perspective.

glennpj replied:
I can try a build with a newer h5py and see how it goes.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -193,39 +212,54 @@ class PyTensorflow(Package, CudaPackage):
+    depends_on('py-wrapt@1.11.0:', type=('build', 'run'), when='@2.7:')
+    depends_on('py-wrapt@1.12.1:1.12', type=('build', 'run'), when='@2.4:2.6')
     depends_on('py-wrapt@1.11.1:', type=('build', 'run'), when='@1.12.1,1.14:2.3')
+    depends_on('py-wheel@0.32.0:', type=('build', 'run'), when='@2.7:')
```
```suggestion
    depends_on('py-wheel@0.32:0', type=('build', 'run'), when='@2.7:')
```

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -138,38 +144,51 @@ class PyTensorflow(Package, CudaPackage):
     depends_on('py-enum34@1.1.6:', type=('build', 'run'), when='@1.5: ^python@:3.3')
     depends_on('py-enum34@1.1.6:', type=('build', 'run'), when='@1.4.0:1.4.1')
-    depends_on('py-gast@0.3.3', type=('build', 'run'), when='@2.2:')
+    depends_on('py-gast@0.2.1:0.5.0', type=('build', 'run'), when='@2.7:')
```
```suggestion
    depends_on('py-gast@0.2.1:0.4', type=('build', 'run'), when='@2.7:')
```

glennpj replied:
Sorry, I misread that line in the file as "<=" but it is just "<".

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -193,39 +212,54 @@ class PyTensorflow(Package, CudaPackage):
-    depends_on('py-wrapt@1.12.1:1.12', type=('build', 'run'), when='@2.4.0:')
+    depends_on('py-tensorboard@2.5.0:2.5', type=('build', 'run'), when='@2.5')
+    depends_on('py-tensorboard@2.6.0:2.6', type=('build', 'run'), when='@2.6')
+    depends_on('py-tensorboard@2.7.0:2.7', type=('build', 'run'), when='@2.7')
```
`setup.py` says `~= 2.6` which translates to `2.6:2` in Spack, I'm guessing the other versions are wrong too

adamjstewart replied:
You've added a tensorboard dep that we've previously removed due to a circular dependency, but we're still commenting out the tensorboard dep in `setup.py` in `post_configure_fixes`, should we add a version check to guard that? Also, the `setup.py` also lists `keras` and `tensorflow-io-gcs-filesystem` as deps. If those have circular deps, you can try commenting them out. But if we don't install them, setuptools/pip will, which we don't want.

glennpj replied:
> `setup.py` says `~= 2.6` which translates to `2.6:2` in Spack, I'm guessing the other versions are wrong too

I understand that, but my understanding is that the versions need to match to the minor version. The `@2.6:2` spec would allow tensorboard-2.7 to be used with tensorflow-2.6, which I believe will cause issues.

adamjstewart replied:
Gotcha, might want to submit a PR to the TF devs to use `~= 2.6.0` or `== 2.6.*` instead.

glennpj replied:
I just committed a change to adjust the dependencies listed in the setup.py file for 

- tensorboard
- tensorflow-estimator
- keras
- tensorflow-io-gcs-filesystem

The last one is new for tensorflow-2.7 and I think would best be added in a separate PR.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -193,39 +212,54 @@ class PyTensorflow(Package, CudaPackage):
-        depends_on('py-grpcio@1.32.0:1.32', type=('build', 'run'), when='@2.4:')
+        depends_on('py-grpcio@1.32.0:1.32', type=('build', 'run'), when='@2.4.0:2.4')
+        depends_on('py-grpcio@1.34.0:1.34', type=('build', 'run'), when='@2.5.0:2.5')
+        depends_on('py-grpcio@1.37.0:1', type=('build', 'run'), when='@2.6.0:')
```
Latest version wants `1.24.3:1`

glennpj committed:
Update var/spack/repos/builtin/packages/py-tensorflow/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

glennpj committed:
Update var/spack/repos/builtin/packages/py-tensorflow/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

glennpj committed:
Fix py-grpcio dependency for version 2.7

Also, make sure py-h5py mpi specs are consistent.

glennpj committed:
Add llvm as run dependency.

glennpj committed:
Fix python spec for py-tensorboard

glennpj committed:
Fix py-google-auth spec for py-tensorboard

glennpj committed:
Do not override the pip spec for tensorboard-plugin-wit

glennpj committed:
Converted py-tensorboard-plugin-wit to wheel only package

glennpj committed:
Fix bazel dependency spec in tensorflow

glennpj committed:
Adjust pip masks

- allow tensorboard to be specified in pip constraints
- mask tensorflow-estimator

glennpj committed:
Remove blank line at end of file

glennpj commented:
@adamjstewart and @aweits 

I converted the py-tensorboard-plugin-wit package to a wheel only package. Given the comments that I read on the project page that seemed like the best way to go. Let me know if that might be problematic and I will revert it back.

glennpj committed:
Adjust pip constraints in setup.py

Also, adjust constraint on a patch that is fixed in 2.7

glennpj committed:
Fix flake8 error

Adjust formatting for consistency.

glennpj commented:
Remaining items:

- [x] create py-libclang package for tensorflow-2.7
py-clang is needed for tensorflow-2.6
- [x] check build with newer h5py version

glennpj committed:
Get bazel dep right

glennpj committed:
Fix old cudnn dependency, caught in audit test

glennpj committed:
Adjust the regex to ensure proper line is changed

glennpj committed:
Add py-libclang package

- Stripped the py-clang package down to just version 5
- added comments to indicate the purpose of py-clang and that
  py-libclang should be preferred
- set dependencies accordingly in py-tensorflow

glennpj committed:
Remove cap on py-h5py dependency for v2.7

glennpj committed:
Add TODO entries for tensorflow-io-gcs-filesystem

glennpj committed:
Edit some comments

glennpj commented:
- I believe I have addressed all of the comments and requests. I created a py-libclang package for tensorflow-2.7 and greater. I stripped down the py-clang package to just what is needed for tensorflow-2.6 and commented it. I removed the llvm dependency from py-clang to prevent having that force an older GCC for tensorflow-2.6. That is not ideal but seemed like a good way to meet dependency specs and still allow people to use the bindings if they need them. See 
- https://github.com/spack/spack/pull/27138#discussion_r777248087
- https://github.com/spack/spack/pull/27138#discussion_r777493306
- https://github.com/spack/spack/pull/27138#discussion_r777580062

for reference.

adamjstewart reviewed (commented):
Can you rebase and update the packages to use the new `pip install` method I use in other TF packages?

glennpj commented:
I rebased this and am verifying that it builds. It does not, as bazel-4 now fails.
The first error is:
```console
10 errors found in build log:
     11       
     12       
     13       Loading: 0 packages loaded
     14           Fetching @bazel_toolchains; fetching
     15       
     16       
  >> 17       DEBUG: /tmp/bazel_h1CNr2oa/out/external/bazel_toolchains/rules/rbe_repo/version_check.bzl:59:14:
     18       Current running Bazel is not a release version and one was not defined explicitly in rbe_autoconfig target. Falling back to '
              3.1.0'
     19       Loading: 0 packages loaded
```
which is, puzzling.

glennpj commented:
Well, bazel-4.0.0 does not build with gcc-11. I guess I was using gcc-10 before. There is a patch on the bazel site so I will create a separate PR for that.

glennpj committed:
Add phases and select python in PATH for tensorboard-data-server

glennpj commented:
The bazel-4/gcc-11 issue is more complicated than it appeared at first. I guess that should be no surprise. I went back to gcc-10 to finish testing this PR after rebasing so it should be ready now.

adamjstewart commented on var/spack/repos/builtin/packages/py-clang/package.py:
```diff
@@ -0,0 +1,26 @@
+        "5.0", sha256="ceccae97eda0225a5b44d42ffd61102e248325c2865ca53e4407746464a5333a"
+    )
+
+    depends_on("python@2.7:2.8,3:", type=("build", "run"))
```
```suggestion
    depends_on("python@2.7:", type=("build", "run"))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-libclang/package.py:
```diff
@@ -0,0 +1,50 @@
+
+    depends_on("python@2.7:2.8,3.3:", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
+    depends_on("py-wheel", type=("build", "run"))
```
I don't think wheel is needed at run-time. `PythonPackage` already adds an implicit build dep on wheel, so you could probably remove this.

adamjstewart commented on var/spack/repos/builtin/packages/py-clang/package.py:
```diff
@@ -0,0 +1,26 @@
+    # py-libclang package should be used instead. It not recommended to add
+    # updated versions to this package. Note that no llvm dependency is
+    # specified here because llvm-5 will cause spack concretization to use an
+    # equally old GCC.
```
Will this package work if no LLVM is installed on the system anywhere?

glennpj replied:
It is a shame that rebasing causes comments to be lost. There was an unresolved conversation around this where I had an open question about how to handle this dependency. There are two things that could be done.

1. leave the dependency in place so people could use the bindings if they had llvm-5 available.
2. remove this dependency for this version of TF and mask it in the setup.py file.

I chose to do (1) for now simply because the package was already in the PR and it would be easier to flip (1) to (2) than it would be to flip (2) to (1).

adamjstewart replied:
Could we keep LLVM as a dep? This only affects one specific version of TF, right?

glennpj replied:
Yes, but as I tried to convey in my comment, that will then cause the DAG to have a really old compiler version. It seems that would cause problems for people as well. At least at my site, it would make TF-2.6 undeployable, particularly since it involves python.

glennpj replied:
My preferred path is option (2) above. Like I said, it was an open question and I structured it to be flexible. The clang dependency was new in TF-2.6 and changed to the other clang project in TF-2.7. As far as I know, the clang bindings are not strictly needed and for those who need them, TF-2.7 is available.

glennpj replied:
> Could we keep LLVM as a dep? This only affects one specific version of TF, right?

All of 2.6, so 2.6.0, 2.6.1, 2.6.2.

adamjstewart replied:
I would personally add the dep on LLVM 5 if `py-clang@5` does indeed only work for LLVM 5 and let the concretizer deal with it. But I don't plan on using old versions of TF so I'll leave the final decision to you

glennpj replied:
I removed, and masked, the py-clang dependency for TF-2.6. That is what felt best to me, especially since TF-2.7 is using the newer py-libclang package.

adamjstewart commented on var/spack/repos/builtin/packages/py-libclang/package.py:
```diff
@@ -0,0 +1,50 @@
+    def patch(self):
+        filter_file(
+            r"source_dir = './native/'",
+            r"source_dir = '{0}'".format(self.spec["llvm"].prefix.lib),
```
Should this be `spec['llvm'].libs.directories[0]`? Not all OSes use `prefix.lib`, some use `prefix.lib64`

glennpj replied:
That would be better but I get the following error:
```console
==> Error: NoLibrariesError: Unable to recursively locate llvm libraries in /opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-cascadelake/gcc-10.3.0/llvm-11.1.0-dvkyuwd3s3524vkpkrme56zf6aft55ud
```

adamjstewart replied:
Just add this to the LLVM package:
```python
@property
def libs(self):
    shared = '+shared_libs' in self.spec
    return find_libraries('libc++', self.prefix, shared=shared, recursive=True)
```
or something like that.

glennpj replied:
I think this has to with the fact that there is no library named, `libllvm.so` in the llvm lib directory.

adamjstewart replied:
Correct, if the library name is different than the package name, you need to add a custom `libs` attribute. `'libc++'` might not actually be the best name since I think that is only available if you build with `+libcxx`. Are there any libraries that always get installed?

adamjstewart commented on var/spack/repos/builtin/packages/py-libclang/package.py:
```diff
@@ -0,0 +1,50 @@
+
+    def patch(self):
+        filter_file(
+            r"source_dir = './native/'",
```
No need for `r` on these strings, there are no backslashes

glennpj replied:
Force of habit.

adamjstewart commented on var/spack/repos/builtin/packages/py-libclang/package.py:
```diff
@@ -0,0 +1,50 @@
+        filter_file(
+            r"source_dir = './native/'",
+            r"source_dir = '{0}'".format(self.spec["llvm"].prefix.lib),
+            "setup.py",
```
```suggestion
            "setup.py", string=True,
```

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard-data-server/package.py:
```diff
@@ -0,0 +1,52 @@
+    """Fast data loading for TensorBoard"""
+
+    homepage = "https://github.com/tensorflow/tensorboard/tree/master/tensorboard/data/server"
+    git      = "https://github.com/tensorflow/tensorboard"
```
Can we download from URL so that we can checksum? Or does tensorboard-data-server versioning not match up with tensorboard versioning?

glennpj replied:
The data-server code is versioned differently and set at specific commits.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard-data-server/package.py:
```diff
@@ -0,0 +1,52 @@
+
+    depends_on('python@3.6:', type=('build', 'run'))
+    depends_on('py-setuptools', type='build')
+    depends_on('py-wheel', type='build')
+    depends_on('py-pip', type='build')
```
`PythonPackage` already adds wheel/pip build deps, these can be removed

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard-data-server/package.py:
```diff
@@ -0,0 +1,52 @@
+    depends_on('py-pip', type='build')
+    depends_on('rust', type='build')
+
+    phases = ['build', 'install']
```
Instead of overriding phases I would just use `@run_before('install')` to decorate `build`. That way we don't lose things like post-install tests.

glennpj replied:
I think phases makes more sense. How would using phases break post-install tests?

adamjstewart replied:
The post-install tests are registered to the phase when the base class is initialized. If you override `phases`, all tests are deregistered and you have to add them again. @alalazo is this understanding correct?

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard-data-server/package.py:
```diff
@@ -0,0 +1,52 @@
+
+        with working_dir(join_path('tensorboard', 'data', 'server',
+                                   'pip_package')):
+            python = which('python')
```
This is already defined in the `python` package, you can remove this line

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard-plugin-wit/package.py:
```diff
@@ -3,7 +3,7 @@
+from spack import *
 
 
 class PyTensorboardPluginWit(Package):
```
If this PR takes much longer, I would be interested in splitting this into a separate PR and merging immediately. I'm having build issues with this package on multiple platforms.

adamjstewart replied:
This should subclass `PythonPackage` now, we now support installing wheels. Then you can remove `extends` and the pip dep

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard/package.py:
```diff
@@ -32,22 +37,34 @@ class PyTensorboard(Package):
+
+    # py-tensorboard-plugin-wit only builds with versions of bazel older than
+    # what is needed for py-tensorboard-2.5 and greater. Work around that for
+    # now by using a wheel.
```
This comment prob belongs in the `py-tensorboard-plugin-wit` package. I encountered the same issue with `py-jupyter-packaging` and ended up splitting it into multiple packages for each version, but I'm also fine with the wheel solution.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard/package.py:
```diff
@@ -32,22 +37,34 @@ class PyTensorboard(Package):
 
     phases = ['configure', 'build', 'install']
 
+    # Version 2.6.0 does not build in parallel
```
Only version 2.6.0 or also all older versions?

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard/package.py:
```diff
@@ -32,22 +37,34 @@ class PyTensorboard(Package):
+    # now by using a wheel.
+    depends_on('py-tensorboard-plugin-wit@1.6.0:', type=('build', 'run'))
+
+    depends_on('py-tensorboard-data-server@0.6.0:0.6', type=('build', 'run'), when='@2.5:')
```
```suggestion
    depends_on('py-tensorboard-data-server@0.6', type=('build', 'run'), when='@2.5:')
```
This should also work?

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard/package.py:
```diff
@@ -16,13 +16,18 @@ class PyTensorboard(Package):
     depends_on('bazel@2.1.0:', type='build', when='@2.2.0:')
+    depends_on('bazel@3.7.0:', type='build', when='@2.5.0:')
     depends_on('py-pip', type='build')
     depends_on('py-wheel', type='build')
```
0.26+, at least in the latest version

glennpj replied:
Ah, I removed the wrong `depends_on('py-wheel')` line when I resolved the rebase conflicts.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -140,38 +146,51 @@ class PyTensorflow(Package, CudaPackage):
     depends_on('py-astor@0.6.0:', type=('build', 'run'), when='@1.6:2.1')
     depends_on('py-backports-weakref@1.0:', type=('build', 'run'), when='@1.3: ^python@:3.3')
     depends_on('py-backports-weakref@1.0rc1', type=('build', 'run'), when='@1.2.0:1.2.1')
+    depends_on('py-clang@5.0.0:5.0', type=('build', 'run'), when='@2.6')
```
```suggestion
    depends_on('py-clang@5.0', type=('build', 'run'), when='@2.6')
```
This should work?

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -195,45 +214,63 @@ class PyTensorflow(Package, CudaPackage):
-    depends_on('flatbuffers+python@1.12.0:1.12', type=('build', 'run'), when='@2.4.0:')
+    depends_on('protobuf@:3.12', when='@:2.4')
+    depends_on('protobuf@:3.17')
+    depends_on('flatbuffers+python@1.12:2.9', type=('build', 'run'), when='@2.7:')
```
```suggestion
    depends_on('flatbuffers+python@1.12:2', type=('build', 'run'), when='@2.7:')
```
2.10 would also work

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -195,45 +214,63 @@ class PyTensorflow(Package, CudaPackage):
+    depends_on('protobuf@:3.12', when='@:2.4')
+    depends_on('protobuf@:3.17')
+    depends_on('flatbuffers+python@1.12:2.9', type=('build', 'run'), when='@2.7:')
+    depends_on('flatbuffers+python@1.12.0:1.12', type=('build', 'run'), when='@2.4.0:2.6')
```
```suggestion
    depends_on('flatbuffers+python@1.12', type=('build', 'run'), when='@2.4.0:2.6')
```
This should work?

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -140,38 +146,51 @@ class PyTensorflow(Package, CudaPackage):
     # propagate the mpi variant setting for h5py/hdf5 to avoid unexpected crashes
-    depends_on('py-h5py@2.10.0:2.10+mpi', type=('build', 'run'), when='@2.2:+mpi')
-    depends_on('py-h5py@2.10.0:2.10~mpi', type=('build', 'run'), when='@2.2:~mpi')
+    depends_on('py-h5py@2.9.0:', type=('build', 'run'), when='@2.7:+mpi')
+    depends_on('py-h5py~mpi@2.9.0:', type=('build', 'run'), when='@2.7:~mpi')
```
So `py-tensorflow+mpi` can be built with either `py-h5py+mpi` or `py-h5py~mpi`, and `py-tensorflow~mpi` can only be built with `py-h5py~mpi`, is that correct?

glennpj replied:
Hmm, I can not remember why I dropped the `+mpi`. It is entirely possible that I lost it a some point during copy/paste/edit operations. I will add that back in to get the pattern that was there originally.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -140,38 +146,51 @@ class PyTensorflow(Package, CudaPackage):
+    depends_on('py-h5py~mpi@2.9.0:', type=('build', 'run'), when='@2.7:~mpi')
+    depends_on('py-h5py@3.1.0:3.1', type=('build', 'run'), when='@2.5:2.6+mpi')
+    depends_on('py-h5py~mpi@3.1.0:3.1', type=('build', 'run'), when='@2.5:2.6~mpi')
+    depends_on('py-h5py@2.10.0:2.10', type=('build', 'run'), when='@2.2:2.4.999+mpi')
+    depends_on('py-h5py~mpi@2.10.0:2.10', type=('build', 'run'), when='@2.2:2.4.999~mpi')
```
```suggestion
    depends_on('py-h5py@2.10', type=('build', 'run'), when='@2.2:2.4+mpi')
    depends_on('py-h5py~mpi@2.10', type=('build', 'run'), when='@2.2:2.4~mpi')
```

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -140,38 +146,51 @@ class PyTensorflow(Package, CudaPackage):
+    depends_on('py-h5py~mpi@3.1.0:3.1', type=('build', 'run'), when='@2.5:2.6~mpi')
+    depends_on('py-h5py@2.10.0:2.10', type=('build', 'run'), when='@2.2:2.4.999+mpi')
+    depends_on('py-h5py~mpi@2.10.0:2.10', type=('build', 'run'), when='@2.2:2.4.999~mpi')
     depends_on('hdf5+mpi', type='build', when='@2.2:+mpi')
     depends_on('hdf5~mpi', type='build', when='@2.2:~mpi')
```
Does TF actually link to hdf5 or does it just use `py-h5py` for everything?

glennpj replied:
TF does not link to HDF5 libraries. I will note that those specs were already there but my guess is they are specified in order to propagate the `mpi` variant. Both the `hdf5` and `py-h5py` packages have `+mpi` set by default, and the `py-h5py` package only sets the `+mpi` variant for `hdf5`, not the `~mpi` variant. So, MPI would be pulled in regardless of the `mpi` variant in TF if the `hdf5` specs were not there. Setting the `hdf5` variants explicitly allows `mpi` to be toggled and keep the `mpi` variant in sync across the packages. Figuring out what variants work together is something that could be further explored, but I do not think this PR is the correct vehicle for that.

adamjstewart reviewed (commented):
Apologies if any of these things were already discussed, rebasing can delete comments from the history.

I would love to make these kinds of changes directly to the package myself, but my bazel installation is still broken for some reason so I can't test anything. If someone can fix bazel for me I would be more than happy to finish up this PR.

The TF package is becoming massive. We should consider deprecating older versions of TF to make it easier to maintain. I'm sure a lot of people are still stuck on TF 1, but there's no reason we need to keep TF 0 around.

glennpj committed:
py-libclang

- remove py-wheel dependency
- remove raw string notation in filter_file

glennpj committed:
py-tensorboard-data-server

- remove py-wheel dep
- remove py-pip dep
- use python from package class

glennpj committed:
py-tensorboard-plugin-wit

- switch to PythonPackage
- add version 1.8.1
- remove unneeded code

glennpj committed:
Add comment as to why a wheel is need for tensorboard-plugin-wit

glennpj committed:
remove which pip from tensorboard-data-server

glennpj committed:
Fix dependency specs in tensorboard

glennpj committed:
tweak dependencies for tensorflow

glennpj committed:
fix python constraint

glennpj commented:
I pushed some commits that address most of the issues brought up. The py-clang dependency for TF-2.6 still needs to be addressed.

glennpj commented:
This PR now depends on #28621.

glennpj committed:
Use llvm libs property

glennpj committed:
py-tensorboard-data-server

- merge build into install
- use std_pip_args

adamjstewart commented on var/spack/repos/builtin/packages/py-clang/package.py:
```diff
@@ -0,0 +1,26 @@
+    """Python bindings for clang from clang release branches"""
+
+    # This package is needed for py-tensorflow@2.6. Generally speaking the
+    # py-libclang package should be used instead. It not recommended to add
```
```suggestion
    # py-libclang package should be used instead. It is not recommended to add
```

glennpj replied:
This package has been removed.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard-data-server/package.py:
```diff
@@ -0,0 +1,45 @@
+    def setup_build_environment(self, env):
+        env.set('CARGO_HOME', self.stage.source_path)
+
+    def install(self, spec, prefix):
```
If we do:
```suggestion
    @run_before('install')
    def build(self):
```
Then we can remove the pip/wheel stuff below. This is optional, just a suggestion.

glennpj replied:
This package does not install correctly with the install method removed. I think it has to do with building with cargo.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard-plugin-wit/package.py:
```diff
@@ -3,10 +3,10 @@
 
 
-class PyTensorboardPluginWit(Package):
+class PyTensorboardPluginWit(PythonPackage):
```
Changes to this package can be removed now

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorboard/package.py:
```diff
@@ -16,38 +16,52 @@ class PyTensorboard(Package):
     depends_on('py-google-auth-oauthlib@0.4.1:0.4', type=('build', 'run'))
-    depends_on('py-tensorboard-plugin-wit@1.6.0:', type=('build', 'run'), when='@2.2.0:')
+
+    depends_on('py-tensorboard-plugin-wit@1.6.0:', type=('build', 'run'))
```
Is this needed for all versions now?

glennpj replied:
I removed the `when` attribute because version 2.2.0 is the oldest version in the package recipe and seemed unnecessary.

adamjstewart commented on var/spack/repos/builtin/packages/py-tensorflow/package.py:
```diff
@@ -195,45 +214,63 @@ class PyTensorflow(Package, CudaPackage):
-    depends_on('protobuf')
-    depends_on('flatbuffers+python@1.12.0:1.12', type=('build', 'run'), when='@2.4.0:')
+    depends_on('protobuf@:3.12', when='@:2.4')
+    depends_on('protobuf@:3.17')
```
Does this fix #28612? If so, can you add "fixes #28612" to the initial PR description so it gets automatically closed when this is merged?

glennpj replied:
I am pretty sure it does.

adamjstewart reviewed (commented):
This is getting pretty close! Let's try to finish and merge this as soon as the LLVM libs PR is merged. We can fix any remaining issues after this is merged so it will be easier to review.

glennpj committed:
remove py-clang dependency

glennpj committed:
Merge branch 'develop' into py-tensorflow

glennpj committed:
remove my edits to py-tensorboard-plugin-wit

glennpj commented:
> This is getting pretty close! Let's try to finish and merge this as soon as the LLVM libs PR is merged. We can fix any remaining issues after this is merged so it will be easier to review.

I think I am done here.

asesidaa commented:
Hi, since #28621 has been merged, I think it’s time to get this merged and maybe start testing other related pull requests like #27991

adamjstewart commented:
@glennpj can you rebase?

glennpj commented:
> @glennpj can you rebase?

I am not seeing any message in github that indicates a rebase is needed.

adamjstewart commented:
I guess a rebase is only needed if someone wants to actually test this PR and try to build TF, it isn't needed to merge. I'll try to review this afternoon and get these 3 PRs merged.

aweits commented:
> I guess a rebase is only needed if someone wants to actually test this PR and try to build TF, it isn't needed to merge. I'll try to review this afternoon and get these 3 PRs merged.

Let's just merge it and then we can iterate on any issues and/or 2.8.0

adamjstewart reviewed (approved):
Thanks for all the hard work on this, and sorry it took so long to get merged! I can't properly test this at the moment because the previous TF package didn't build correctly on my system either, but I'll try this new version and open an issue if I encounter any problems.
